File: examples/javascript/webhook-handler.js
/**
 * ApiPay.kz - Webhook Handler Example
 *
 * This example demonstrates how to:
 * 1. Receive webhook notifications
 * 2. Verify the signature
 * 3. Handle payment events (invoices and subscriptions)
 *
 * Usage:
 *   WEBHOOK_SECRET=your_secret PORT=3000 node webhook-handler.js
 *
 * Test with:
 *   curl -X POST http://localhost:3000/webhook \
 *     -H "Content-Type: application/json" \
 *     -H "X-Webhook-Signature: sha256=..." \
 *     -d '{"event":"invoice.status_changed","invoice":{"id":42,"status":"paid"},"source":"api"}'
 */

const http = require('http')
const crypto = require('crypto')

const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET
const PORT = process.env.PORT || 3000

/**
 * Verify webhook signature using HMAC-SHA256
 */
function verifySignature(payload, signature, secret) {
  const expected = 'sha256=' + crypto
    .createHmac('sha256', secret)
    .update(payload)
    .digest('hex')

  try {
    return crypto.timingSafeEqual(Buffer.from(expected), Buffer.from(signature))
  } catch {
    return false
  }
}

/**
 * Handle invoice status change event
 */
function handleInvoiceStatusChanged(invoice) {
  console.log(`\nInvoice #${invoice.id} status changed to: ${invoice.status}`)

  switch (invoice.status) {
    case 'paid':
      console.log(`  Payment received!`)
      console.log(`  Amount: ${invoice.amount} KZT`)
      console.log(`  Client: ${invoice.client_name}`)
      console.log(`  Paid at: ${invoice.paid_at}`)
      if (invoice.external_order_id) {
        console.log(`  Order ID: ${invoice.external_order_id}`)
        // TODO: Fulfill the order
      }
      break

    case 'cancelled':
      console.log(`  Invoice was cancelled`)
      break

    case 'expired':
      console.log(`  Invoice expired`)
      break

    default:
      console.log(`  Status: ${invoice.status}`)
  }
}

/**
 * Handle invoice refunded event
 */
function handleInvoiceRefunded(invoice) {
  console.log(`\nInvoice #${invoice.id} refunded`)
  console.log(`  Status: ${invoice.status}`)
  console.log(`  Total refunded: ${invoice.total_refunded} KZT`)
}

/**
 * Handle subscription events
 */
function handleSubscriptionEvent(eventType, data) {
  const sub = data.subscription
  console.log(`\nSubscription #${sub.id} — ${eventType}`)

  switch (eventType) {
    case 'subscription.payment_succeeded':
      console.log(`  Payment succeeded! Invoice #${data.invoice.id}: ${data.invoice.amount} KZT`)
      break

    case 'subscription.payment_failed':
      console.log(`  Payment failed: ${data.reason}`)
      break

    case 'subscription.grace_period_started':
      console.log(`  Grace period: ${sub.grace_period_days} days, ${sub.retry_attempts_remaining} retries left`)
      break

    case 'subscription.expired':
      console.log(`  Subscription expired`)
      break
  }
}

/**
 * HTTP request handler
 */
function handleRequest(req, res) {
  if (req.method !== 'POST' || req.url !== '/webhook') {
    res.writeHead(404)
    res.end('Not Found')
    return
  }

  let body = ''

  req.on('data', chunk => {
    body += chunk.toString()
  })

  req.on('end', () => {
    const signature = req.headers['x-webhook-signature']

    if (!signature) {
      res.writeHead(401)
      res.end('Missing signature')
      return
    }

    if (!verifySignature(body, signature, WEBHOOK_SECRET)) {
      res.writeHead(401)
      res.end('Invalid signature')
      return
    }

    try {
      const event = JSON.parse(body)
      console.log(`\nReceived event: ${event.event} (source: ${event.source})`)
      console.log(`Timestamp: ${event.timestamp}`)

      switch (event.event) {
        case 'invoice.status_changed':
          handleInvoiceStatusChanged(event.invoice)
          break

        case 'invoice.refunded':
          handleInvoiceRefunded(event.invoice)
          break

        case 'subscription.payment_succeeded':
        case 'subscription.payment_failed':
        case 'subscription.grace_period_started':
        case 'subscription.expired':
          handleSubscriptionEvent(event.event, event.data)
          break

        default:
          console.log(`Unknown event type: ${event.event}`)
      }

      res.writeHead(200)
      res.end('OK')

    } catch (error) {
      console.error('Error parsing webhook:', error.message)
      res.writeHead(400)
      res.end('Invalid JSON')
    }
  })
}

// Main
if (!WEBHOOK_SECRET) {
  console.error('Error: WEBHOOK_SECRET environment variable is required')
  console.error('Usage: WEBHOOK_SECRET=your_secret node webhook-handler.js')
  process.exit(1)
}

const server = http.createServer(handleRequest)

server.listen(PORT, () => {
  console.log(`Webhook server listening on port ${PORT}`)
  console.log(`Endpoint: POST http://localhost:${PORT}/webhook`)
  console.log('\nWaiting for webhook events...')
})

File: examples/python/webhook_handler.py
#!/usr/bin/env python3
"""
ApiPay.kz - Webhook Handler Example

This example demonstrates how to:
1. Receive webhook notifications
2. Verify the signature
3. Handle payment events (invoices and subscriptions)

Usage:
    WEBHOOK_SECRET=your_secret python webhook_handler.py

Requirements:
    pip install flask
"""

import hashlib
import hmac
import os
import sys

from flask import Flask, request, abort

app = Flask(__name__)

WEBHOOK_SECRET = os.environ.get('WEBHOOK_SECRET')


def verify_signature(payload: bytes, signature: str, secret: str) -> bool:
    """Verify webhook signature using HMAC-SHA256."""
    expected = 'sha256=' + hmac.new(
        secret.encode(),
        payload,
        hashlib.sha256
    ).hexdigest()

    return hmac.compare_digest(expected, signature)


def handle_invoice_status_changed(invoice: dict):
    """Handle invoice status change event."""
    print(f"\nInvoice #{invoice['id']} status changed to: {invoice['status']}")

    status = invoice['status']

    if status == 'paid':
        print('  Payment received!')
        print(f"  Amount: {invoice.get('amount')} KZT")
        print(f"  Client: {invoice.get('client_name')}")
        if invoice.get('external_order_id'):
            print(f"  Order ID: {invoice['external_order_id']}")
            # TODO: Fulfill the order

    elif status == 'cancelled':
        print('  Invoice was cancelled')

    elif status == 'expired':
        print('  Invoice expired')


def handle_invoice_refunded(invoice: dict):
    """Handle invoice refunded event."""
    print(f"\nInvoice #{invoice['id']} refunded")
    print(f"  Status: {invoice.get('status')}")
    print(f"  Total refunded: {invoice.get('total_refunded')} KZT")


def handle_subscription_event(event_type: str, data: dict):
    """Handle subscription events."""
    sub = data['subscription']
    print(f"\nSubscription #{sub['id']} — {event_type}")

    if event_type == 'subscription.payment_succeeded':
        inv = data['invoice']
        print(f"  Payment succeeded! Invoice #{inv['id']}: {inv['amount']} KZT")

    elif event_type == 'subscription.payment_failed':
        print(f"  Payment failed: {data.get('reason')}")

    elif event_type == 'subscription.grace_period_started':
        print(f"  Grace period: {sub.get('grace_period_days')} days, {sub.get('retry_attempts_remaining')} retries left")

    elif event_type == 'subscription.expired':
        print('  Subscription expired')


@app.route('/webhook', methods=['POST'])
def webhook():
    """Handle incoming webhook requests."""
    payload = request.get_data()
    signature = request.headers.get('X-Webhook-Signature', '')

    if not signature:
        abort(401)

    if not verify_signature(payload, signature, WEBHOOK_SECRET):
        abort(401)

    try:
        event = request.get_json()
        event_type = event['event']
        print(f"\nReceived event: {event_type} (source: {event.get('source')})")
        print(f"Timestamp: {event.get('timestamp')}")

        if event_type == 'invoice.status_changed':
            handle_invoice_status_changed(event['invoice'])

        elif event_type == 'invoice.refunded':
            handle_invoice_refunded(event['invoice'])

        elif event_type.startswith('subscription.'):
            handle_subscription_event(event_type, event['data'])

        else:
            print(f"Unknown event type: {event_type}")

        return 'OK', 200

    except Exception as e:
        print(f'Error parsing webhook: {e}')
        abort(400)


def main():
    if not WEBHOOK_SECRET:
        print('Error: WEBHOOK_SECRET environment variable is required')
        print('Usage: WEBHOOK_SECRET=your_secret python webhook_handler.py')
        sys.exit(1)

    port = int(os.environ.get('PORT', 5000))
    print(f'Webhook server listening on port {port}')
    print(f'Endpoint: POST http://localhost:{port}/webhook')
    print('\nWaiting for webhook events...')

    app.run(host='0.0.0.0', port=port)


if __name__ == '__main__':
    main()
